File: voigtinference/bench/crosscheck.jl
# Check VoigtInference.jl against the Python reference values.
#
#   python bench/crosscheck.py
#   julia -t 1 --project=. /path/to/bench/crosscheck.jl
#
# Reports the maximum relative difference per quantity.  The grid spans the
# bulk, the neighbourhood of the far-tail switch, and several decades beyond,
# so a disagreement confined to the tail branch shows up rather than being
# averaged away.

using VoigtInference
using Printf

# Agreement expected between the two implementations, per quantity.
#
# All 1e-12, and in practice every entry comes back exactly 0.0: the two
# implementations perform identical floating-point operations in identical
# order, so they agree bit for bit in both branches.  Keep it tight -- slack
# here would hide a regression.
#
# Note that agreement is not accuracy.  V(Z|y) in particular is a difference of
# two terms of size sqrt(2/pi)*sigma*gamma/K returning a result of size sigma^2;
# near |ytil| ~ 400*scale those terms are ~7.0e5 and the answer is ~0.25, so it
# carries only ~10 correct digits no matter how it is computed.  That is a
# property of the formula, bounded in tests/test_accuracy.py against
# high-precision truth -- it is not something this file can or should measure.
const TOL = Dict("pdf" => 1e-12, "logpdf" => 1e-12, "score" => 1e-12,
                 "hessian" => 1e-12, "condmean" => 1e-12, "fisher" => 1e-12,
                 "condvar" => 1e-12)
tol_for(name) = get(TOL, name, 1e-12)

function reldiff(a, b)
    return abs(a - b) / max(abs(b), 1e-300)
end

function main()
    path = isempty(ARGS) ? joinpath(@__DIR__, "reference_python.txt") : ARGS[1]
    if !isfile(path)
        error("reference file not found: $path  (run bench/crosscheck.py first)")
    end

    mu = 0.0
    sg = 0.0
    gm = 0.0
    ncases = 0
    npoints = 0
    mle_rows = Vector{Vector{Float64}}()

    worst_d = Dict{String,Float64}()
    worst_y = Dict{String,Float64}()

    function note(name, d, y)
        if d > get(worst_d, name, -1.0)
            worst_d[name] = d
            worst_y[name] = y
        end
        return nothing
    end

    for line in eachline(path)
        if startswith(line, "#")
            continue
        end
        parts = split(strip(line))
        if isempty(parts)
            continue
        end
        tag = parts[1]
        if tag == "theta"
            mu = parse(Float64, parts[2])
            sg = parse(Float64, parts[3])
            gm = parse(Float64, parts[4])
            ncases += 1
        elseif tag == "fisher"
            v = parse.(Float64, parts[2:10])
            info = voigt_fisher(mu, sg, gm; nodes = 400)
            for i in 1:3
                for j in 1:3
                    ref = v[3 * (i - 1) + j]   # written row-major
                    # the mu off-diagonals are zero by symmetry; compare
                    # absolutely there
                    d = abs(ref) < 1e-10 ? abs(info[i, j] - ref) :
                                           reldiff(info[i, j], ref)
                    note("fisher", d, 0.0)
                end
            end
        elseif tag == "point"
            p = parse.(Float64, parts[2:end])
            npoints += 1
            y = p[1]
            note("pdf", reldiff(voigt_pdf(y, mu, sg, gm), p[2]), y)
            note("logpdf", reldiff(voigt_logpdf(y, mu, sg, gm), p[3]), y)

            sc = voigt_score(y, mu, sg, gm)
            for i in 1:3
                note("score", reldiff(sc[i], p[3 + i]), y)
            end

            H = voigt_hessian(y, mu, sg, gm)
            idx = ((1, 1), (1, 2), (1, 3), (2, 2), (2, 3), (3, 3))
            for k in 1:6
                i, j = idx[k]
                note("hessian", reldiff(H[i, j], p[6 + k]), y)
            end

            note("condmean", reldiff(voigt_condmean(y, mu, sg, gm), p[13]), y)
            note("condvar", reldiff(voigt_condvar(y, mu, sg, gm), p[14]), y)
        elseif tag == "mle"
            push!(mle_rows, parse.(Float64, parts[2:end]))
        end
    end

    @printf("%d theta cases;  %d evaluation points\n\n", ncases, npoints)

    @printf("%-12s%14s%14s%16s\n", "quantity", "max rel diff", "tolerance", "at y")
    ok = true
    for name in ("pdf", "logpdf", "score", "hessian", "condmean", "condvar", "fisher")
        if !haskey(worst_d, name)
            continue
        end
        d = worst_d[name]
        y = worst_y[name]
        tol = tol_for(name)
        flag = d <= tol ? "" : "   <-- ABOVE TOLERANCE"
        if d > tol
            ok = false
        end
        @printf("%-12s%14.2e%14.0e%16.6g%s\n", name, d, tol_for(name), y, flag)
    end

    if !isempty(mle_rows)
        println("\nMLE on the shared benchmark samples:")
        @printf("%8s%14s%14s%14s%14s\n", "n", "d(mu)", "d(sigma)", "d(gamma)", "d(loglik)")
        for row in mle_rows
            n = Int(row[1])
            f = joinpath(@__DIR__, "data", "y_$(n).f64")
            if !isfile(f)
                @printf("%8d   (data file missing, skipped)\n", n)
                continue
            end
            yn = collect(reinterpret(Float64, read(f)))
            r = voigt_mle(yn)
            ds = (reldiff(r.μ, row[2]), reldiff(r.σ, row[3]),
                  reldiff(r.γ, row[4]), reldiff(r.loglik, row[5]))
            @printf("%8d%14.2e%14.2e%14.2e%14.2e\n", n, ds...)
            # optimiser-path agreement enforced too (looser than the
            # evaluation tolerance: paths may differ at the level of the
            # convergence tolerance)
            if maximum(ds) > 1e-11
                ok = false
                println("        ^ ABOVE the 1e-11 MLE-path tolerance")
            end
        end
        println("\n(MLE differences reflect the optimiser path, not the formulas;")
        println(" they should sit at the level of the convergence tolerance.)")
    end

    println()
    if ok
        println("PASS: every quantity is within its tolerance.")
    else
        println("FAIL: at least one quantity exceeds the tolerance.")
    end
    exit(ok ? 0 : 1)
end

main()

File: voigtinference/bench/crosscheck.py
"""Write Python reference values for the Julia implementation to check against.

    python bench/crosscheck.py            # writes bench/reference_python.json
    julia --project=. ../bench/crosscheck.jl   # reads it, reports differences

The grid deliberately spans the bulk, the neighbourhood of the far-tail switch,
and several decades beyond it, so that any disagreement in the tail branch shows
up rather than being averaged away.
"""

from __future__ import annotations

import argparse
import json
import pathlib
import sys

import numpy as np

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent.parent / "src"))

from voigtinference import (  # noqa: E402
    voigt_condmean,
    voigt_condvar,
    voigt_fisher,
    voigt_hessian,
    voigt_logpdf,
    voigt_mle,
    voigt_pdf,
    voigt_score,
)

HERE = pathlib.Path(__file__).resolve().parent

# The primary case plus the two width-ratio extremes: gamma/sigma ~ 1e4 puts
# the CENTER of the profile in the Cauchy-limit branch (the regime of the
# regime of the extreme-ratio regression), gamma/sigma ~ 1e-4 approaches the Gaussian
# limit.  The Fisher quadrature is compared for the primary case only (at
# extreme width ratios the two node generators' last-ulp differences are
# amplified by the integrand, which is a property of quadrature, not of the
# formulas this file checks).
THETAS = [
    (0.3, 1.2, 0.7, True),
    (0.0, 1.0, 1.0e4, False),
    (0.0, 1.0, 1.0e-4, False),
]


def grid(mu, sigma, gamma):
    scale = np.sqrt(sigma**2 + gamma**2)
    bulk = scale * np.array(
        [-8.0, -3.0, -1.0, -0.25, 0.0, 0.25, 1.0, 3.0, 8.0, 40.0]
    )
    tail = scale * np.array([3e2, 7e2, 1e3, 1.5e3, 1e4, 1e6, 1e8])
    return np.concatenate([mu + bulk, mu + tail, mu - tail])


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("-o", "--out", default=str(HERE / "reference_python.json"))
    args = ap.parse_args()

    cases = []
    for mu, sigma, gamma, with_fisher in THETAS:
        y = grid(mu, sigma, gamma)
        case = {
            "theta": {"mu": mu, "sigma": sigma, "gamma": gamma},
            "y": y.tolist(),
            "pdf": voigt_pdf(y, mu, sigma, gamma).tolist(),
            "logpdf": voigt_logpdf(y, mu, sigma, gamma).tolist(),
            "score": voigt_score(y, mu, sigma, gamma).tolist(),
            "hessian": voigt_hessian(y, mu, sigma, gamma).tolist(),
            "condmean": voigt_condmean(y, mu, sigma, gamma).tolist(),
            "condvar": voigt_condvar(y, mu, sigma, gamma).tolist(),
        }
        if with_fisher:
            case["fisher"] = voigt_fisher(mu, sigma, gamma, nodes=400).tolist()
        cases.append(case)
    ref = {"cases": cases}

    data_dir = HERE / "data"
    if (data_dir / "manifest.json").exists():
        manifest = json.loads((data_dir / "manifest.json").read_text())
        fits = {}
        for key, name in manifest["files"].items():
            yn = np.fromfile(data_dir / name, dtype="<f8")
            r = voigt_mle(yn)
            fits[key] = {
                "mu": r.mu,
                "sigma": r.sigma,
                "gamma": r.gamma,
                "loglik": r.loglik,
                "se": r.se.tolist(),
                "iterations": r.iterations,
            }
        ref["mle"] = fits

    pathlib.Path(args.out).write_text(json.dumps(ref, indent=2))

    # dependency-free twin for the Julia side (no JSON parser in the stdlib);
    # %.17g round-trips float64 exactly
    def g(x):
        return format(float(x), ".17g")

    lines = [
        "# Voigt reference values written by bench/crosscheck.py",
        "# theta <mu> <sigma> <gamma>   starts a case; the fisher and point",
        "# lines that follow belong to it",
        "# fisher <9 entries, row-major>  (primary case only)",
        "# point <y> <pdf> <logpdf> <s1 s2 s3> <H11 H12 H13 H22 H23 H33> <condmean> <condvar>",
    ]
    for case in cases:
        th = case["theta"]
        lines.append(
            "theta " + " ".join(g(th[k]) for k in ("mu", "sigma", "gamma"))
        )
        if "fisher" in case:
            lines.append(
                "fisher " + " ".join(g(v) for v in np.ravel(case["fisher"]))
            )
        H = np.asarray(case["hessian"])
        for i, yi in enumerate(case["y"]):
            vals = [
                yi,
                case["pdf"][i],
                case["logpdf"][i],
                *case["score"][i],
                H[i, 0, 0], H[i, 0, 1], H[i, 0, 2],
                H[i, 1, 1], H[i, 1, 2], H[i, 2, 2],
                case["condmean"][i],
                case["condvar"][i],
            ]
            lines.append("point " + " ".join(g(v) for v in vals))
    if "mle" in ref:
        lines.append("# mle <n> <mu> <sigma> <gamma> <loglik> <iterations>")
        for key, f in sorted(ref["mle"].items(), key=lambda kv: int(kv[0])):
            lines.append(
                "mle " + " ".join(
                    [key] + [g(f[k]) for k in ("mu", "sigma", "gamma", "loglik")]
                    + [str(f["iterations"])]
                )
            )

    txt_path = pathlib.Path(args.out).with_suffix(".txt")
    txt_path.write_text("\n".join(lines) + "\n")

    npts = sum(len(c["y"]) for c in cases)
    print(f"wrote {args.out}")
    print(f"wrote {txt_path}")
    print(f"  {len(cases)} theta cases, {npts} evaluation points, "
          f"{'with' if 'mle' in ref else 'without'} MLE fits")


if __name__ == "__main__":
    main()
